libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary.hs

```diff
@@ -0,0 +1,128 @@
+    case embedTrip (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) cborTrip prevTxOut of
+      Left err ->
+        expectationFailure $
+          "Expected to deserialize: =======================================================\n"
+            ++ show err
+      Right (curTxOut :: TxOut era) ->
+        curTxOut `shouldBe` upgradeTxOut prevTxOut
+
+specTxCertUpgrade ::
+  forall era.
+  ( EraTxCert (PreviousEra era)
+  , EraTxCert era
+  , Arbitrary (TxCert (PreviousEra era))
+  , HasCallStack
+  ) =>
+  Spec
+specTxCertUpgrade =
+  prop "upgradeTxCert is preserved through serialization" $ \prevTxCert -> do
+    case embedTrip (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) cborTrip prevTxCert of
+      Left err
+        | Right _ <- upgradeTxCert prevTxCert ->
+            -- We expect deserialization to succeed, when upgrade is possible
+            expectationFailure $
+              "Expected to deserialize: =======================================================\n"
+                ++ show err
+        | otherwise -> pure () -- Both upgrade and deserializer fail successfully
+      Right (curTxCert :: TxCert era)
+        | Right upgradedTxCert <- upgradeTxCert prevTxCert ->
+            curTxCert `shouldBe` upgradedTxCert
+        | otherwise -> expectationFailure "Expected upgradeTxCert to succeed"
+
+specTxAuxDataUpgrade ::
+  forall era t.
+  ( EraTxAuxData (PreviousEra era)
+  , EraTxAuxData era
+  , t era ~ TxAuxData era
+  , Memoized t
+  , Eq (RawType t era)
+  , ToExpr (RawType t era)
+  , Arbitrary (TxAuxData (PreviousEra era))
+  , HasCallStack
+  ) =>
+  Spec
+specTxAuxDataUpgrade =
+  prop "upgradeTxAuxData is preserved through serialization" $ \prevTxAuxData -> do
+    case embedTripAnn (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) prevTxAuxData of
+      Left err ->
+        expectationFailure $
+          "Expected to deserialize: =======================================================\n"
+            ++ show err
+      Right (curTxAuxData :: t era) ->
+        -- We need to do all this MemoBytes trickery because underlying bytes and thus the
+        -- equality of the same type will no longer be the same, despite that the value will
+        zipMemoRawType @t @t expectExprEqual curTxAuxData (upgradeTxAuxData prevTxAuxData)
+
+specScriptUpgrade ::
+  forall era.
+  ( EraScript (PreviousEra era)
+  , EraScript era
+  , Arbitrary (Script (PreviousEra era))
+  , HasCallStack
+  ) =>
+  Spec
+specScriptUpgrade =
+  prop "upgradeScript is preserved through serialization" $ \prevScript -> do
+    case embedTripAnn (eraProtVerHigh @(PreviousEra era)) (eraProtVerLow @era) prevScript of
+      Left err ->
+        expectationFailure $
+          "Expected to deserialize: =======================================================\n"
+            ++ show err
+      Right (curScript :: Script era) ->
+        curScript `shouldBe` upgradeScript prevScript
+
+specUpgrade ::
+  forall era t.
```

Comment:
I am not quite sure how to best explain it, but I'll try to add something. In one sentence: otherwise we can't use MemoBytes interface to compare equality on the raw type.

> The true reason why we need to pass `t` as a type argument is because `TxAuxData` is a type family, so we don't know if the final type will be of the kind `Type -> Type` and can be used with MemoBytes, which requires `t` to be of such kind, because it is later applied to `era`.

^ @nc6 Do you think it is that a good enough explanation to add as a comment in the code?

Reply:
Done.